### ISSUE TITLE
@ComponentScan is activated unexpectedly.

### ISSUE THREAD

**Issue by wangxing-git** [spring-projects/spring-boot#17339]
### spring boot version
`2.1.5.RELEASE`
### TestConfig.java
```java
package org.xyattic.boot.calkin.demo.oauth2.server.config;

import org.springframework.boot.autoconfigure.condition.ConditionalOnBean;
import org.springframework.boot.autoconfigure.condition.ConditionalOnClass;
import org.springframework.context.annotation.ComponentScan;
import org.springframework.context.annotation.Configuration;
import org.springframework.context.annotation.Import;
import org.springframework.core.io.Resource;

/**
 * @author wangxing
 * @create 2019/6/28
 */
@Configuration
@Import(TestConfig.Test.class)
@ConditionalOnBean(Resource.class)
public class TestConfig {

    @ConditionalOnClass(Resource.class)
    @ComponentScan("org.xyattic.boot.calkin.demo.oauth2.test")
    class Test {

    }

}
```
The package `org.xyattic.boot.calkin.demo.oauth2.test` by default is not scanned.
When I activated it under certain conditions, I found that `@ComponentScan` was also activated when the condition was not met.
```java
TestConfig:
      Did not match:
         - @ConditionalOnBean (types: org.springframework.core.io.Resource; SearchStrategy: all) did not find any beans of type org.springframework.core.io.Resource (OnBeanCondition)

   TestConfig.Test:
      Did not match:
         - Ancestor org.xyattic.boot.calkin.demo.oauth2.server.config.TestConfig did not match (ConditionEvaluationReport.AncestorsMatchedCondition)
      Matched:
         - @ConditionalOnClass found required class 'org.springframework.core.io.Resource' (OnClassCondition)
```

**Comment by wilkinsona** [comment]
Thanks for the report. The behaviour that you have observed is part of Spring Framework.

In the first example the unexpected behaviour is occurring because Framework does not consistently consider the conditions on `TestConfig` when evaluating the conditions on its `Test` inner-class. When it's deciding whether or not to process the `@ComponentScan` it only considers the conditions on `Test`. Later on, when it's deciding to process any `@Bean` definitions on `Test` it does consider the conditions on `TestConfig`.

The second example avoid the problem of the first example because `@ConditionalOnProperty` it's a different sort of condition to `@ConditionalOnBean`. Condition evaluation is split into two phases, configuration parsing and bean registration. `@ConditionalOnProperty` can be evaluated in the configuration parsing phase, whereas `@ConditionalOnBean` can only be evaluated in the bean registration phase. In the second example, when `TestConfig` is processed during configuration parsing, `@ConditionalOnProperty` does not match so `TestConfig` is skipped along with its inner-classes. By contrast, in the first example `@ConditionalOnBean` is not considered during the configuration parsing phase  so `Test` and its `@ComponentScan` are then processed.

In short, we need to move this to spring-projects/spring-framework so that the Framework team can take a look.
